Some f2py tests in 1.23.0rc1 depend on availability of fortran compiler

A handful of tests (on linux) now depend on the availability of a fortran-compiler.

Following discussion in https://github.com/conda-forge/numpy-feedstock/pull/272, this dependence should either be removed, or the tests made conditional on the availability of a fortran-compiler.

Example:
```
E          INFO: building 'callstatement' extension
E          error: extension 'callstatement' has Fortran sources but no Fortran compiler found
```

Details:
```
=========================== short test summary info ============================
ERROR f2py/tests/test_return_real.py::TestCReturnReal::test_all[t4] - Runtime...
ERROR f2py/tests/test_return_real.py::TestCReturnReal::test_all[t8] - Runtime...
ERROR f2py/tests/test_return_real.py::TestCReturnReal::test_all[s4] - Runtime...
ERROR f2py/tests/test_return_real.py::TestCReturnReal::test_all[s8] - Runtime...
ERROR f2py/tests/test_semicolon_split.py::TestMultiline::test_multiline - Run...
ERROR f2py/tests/test_semicolon_split.py::TestCallstatement::test_callstatement
19897 passed, 536 skipped, 38 xfailed, 5 xpassed, 4 warnings, 6 errors in 264.34s (0:04:24)
```

CC @HaoZeke

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -344,10 +344,6 @@ class F2PyTest:
         # Check compiler availability first
         if not has_c_compiler():
             pytest.skip("No C compiler available")
-        if not has_f77_compiler():
-            pytest.skip("No Fortran 77 compiler available")
-        if not has_f90_compiler():
-            pytest.skip("No Fortran 90 compiler available")
 
         codes = []
         if self.sources:
@@ -357,11 +353,20 @@ class F2PyTest:
 
         needs_f77 = False
         needs_f90 = False
+        needs_pyf = False
         for fn in codes:
             if str(fn).endswith(".f"):
                 needs_f77 = True
             elif str(fn).endswith(".f90"):
                 needs_f90 = True
+            elif str(fn).endswith(".pyf"):
+                needs_pyf = True
+        if needs_f77 and not has_f77_compiler():
+            pytest.skip("No Fortran 77 compiler available")
+        if needs_f90 and not has_f90_compiler():
+            pytest.skip("No Fortran 90 compiler available")
+        if needs_pyf and not (has_f90_compiler() or has_f77_compiler()):
+            pytest.skip("No Fortran compiler available")
 
         # Build the module
         if self.code is not None:

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -344,6 +344,10 @@ class F2PyTest:
         # Check compiler availability first
         if not has_c_compiler():
             pytest.skip("No C compiler available")
+        if not has_f77_compiler():
+            pytest.skip("No Fortran 77 compiler available")
+        if not has_f90_compiler():
+            pytest.skip("No Fortran 90 compiler available")
 
         codes = []
         if self.sources:
@@ -358,10 +362,6 @@ class F2PyTest:
                 needs_f77 = True
             elif str(fn).endswith(".f90"):
                 needs_f90 = True
-        if needs_f77 and not has_f77_compiler():
-            pytest.skip("No Fortran 77 compiler available")
-        if needs_f90 and not has_f90_compiler():
-            pytest.skip("No Fortran 90 compiler available")
 
         # Build the module
         if self.code is not None:

